balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,640 @@
+		// Convert addresses to endpoints.
+		for i, a := range state.ResolverState.Addresses {
+			newEndpoints[i].Attributes = a.BalancerAttributes
+			newEndpoints[i].Addresses = []resolver.Address{a}
+			// We can't remove balancer attributes here since xds packages use
+			// them to store locality metadata.
+		}
+	}
+
+	// Since we have a new set of addresses, we are again at first pass.
+	b.firstPass = true
+	// If multiple endpoints have the same address, they would use the same
+	// subconn because an AddressMap is used to store subconns.
+	// This would result in attempting to connect to the same subconn multiple times
+	// in the same pass. We don't want this, so we ensure each address is present
+	// in only one endpoint before moving further.
+	newEndpoints = deDupAddresses(newEndpoints)
+
+	// Perform the optional shuffling described in gRFC A62. The shuffling will
+	// change the order of endpoints but not touch the order of the addresses
+	// within each endpoint. - A61
+	if cfg.ShuffleAddressList {
+		newEndpoints = append([]resolver.Endpoint{}, newEndpoints...)
+		internal.ShuffleAddressListForTesting.(func(int, func(int, int)))(len(newEndpoints), func(i, j int) { newEndpoints[i], newEndpoints[j] = newEndpoints[j], newEndpoints[i] })
+	}
+
+	if b.state == connectivity.Ready {
+		// If the previous ready subconn exists in new address list,
+		// keep this connection and don't create new subconns.
+		prevAddr := b.addressIndex.currentAddress()
+		b.addressIndex.updateEndpointList(newEndpoints)
+		if b.addressIndex.seekTo(prevAddr) {
+			return nil
+		}
+	} else {
+		b.addressIndex.updateEndpointList(newEndpoints)
+	}
+
+	// Remove old subConns that were not in new address list.
+	oldAddrs := resolver.NewAddressMap()
+	for _, k := range b.subConns.Keys() {
+		oldAddrs.Set(k, true)
+	}
+
+	// Flatten the new endpoint addresses.
+	newAddrs := resolver.NewAddressMap()
+	for _, endpoint := range newEndpoints {
+		for _, addr := range endpoint.Addresses {
+			newAddrs.Set(addr, true)
+		}
+	}
+
+	// Shut them down and remove them.
+	for _, oldAddr := range oldAddrs.Keys() {
+		if _, ok := newAddrs.Get(oldAddr); ok {
+			continue
+		}
+		val, _ := b.subConns.Get(oldAddr)
+		val.(*scData).subConn.Shutdown()
+		b.subConns.Delete(oldAddr)
+	}
+
+	// If its the first resolver update or the balancer was already READY or
+	// or CONNECTING, enter CONNECTING.
+	// We may be in TRANSIENT_FAILURE due to a previous empty address list,
+	// we should still enter CONNECTING because the sticky TF behaviour mentioned
+	// in A62 applied only when the TRANSIENT_FAILURE is reported dur to connectivity
+	// failures.
+	if b.state == connectivity.Ready || b.state == connectivity.Connecting || oldAddrs.Len() == 0 {
+		// Start connection attempt at first address.
+		b.state = connectivity.Connecting
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
```

Comment:
Thought about this a little more and I feel what you have here currently is the way to go since that would be the correct way to handle this when we are currently `Ready`. I'm marking this comment as revolved. Thanks.